src/main/java/com/shapesecurity/shift/es2016/codegen/TokenStream.java

```diff
@@ -67,17 +67,22 @@ public void put(@Nonnull String tokenStr) {
             if (String.valueOf(tokenStr).equals(".")) {
                 this.writer.append(numberNeedsDoubleDot(this.lastNumber) ? ".." : ".");
                 this.lastNumber = null;
-                this.lastChar = '.';
+                this.lastCodePoint = '.';
                 return;
             }
         }
         this.lastNumber = null;
-        char rightChar = tokenStr.charAt(0);
-        char lastChar = this.lastChar;
-        this.lastChar = tokenStr.charAt(tokenStr.length() - 1);
-        if ((lastChar == '+' || lastChar == '-') && lastChar == rightChar ||
-                Utils.isIdentifierPart(lastChar) && Utils.isIdentifierPart(rightChar) ||
-                lastChar == '/' && (rightChar == 'i' || rightChar == '/')) {
+        int rightCodePoint = tokenStr.codePointAt(0);
+        int lastCodePoint = this.lastCodePoint;
+        char lastChar = tokenStr.charAt(tokenStr.length() - 1);
+        if (lastChar >= 0xDC00 && lastChar <= 0xDFFF) {
```

Comment:
We should probably make a helper in Utils like `isTrailSurrogate` (and `isLeadSurrogate`) and use it here and wherever else we perform this test -- probably at least the parser.